cloudinit/cmd/status.py

```diff
@@ -430,6 +431,15 @@ def get_latest_event(status_v1):
     return latest_event
 
 
+def get_not_started_errors(status_v1) -> List[str]:
+    """Return a list of errors from status_v1 that are not started."""
+    return [
+        f"Failed to start stage: {stage_name}"
+        for stage_name, stage_info in status_v1.items()
+        if stage_name in STAGE_NAME.keys() and stage_info.get("start") is None
+    ]
```

Comment:
Couldn't this logic instead be iterating by known STAGE_NAMES instead of looping over status_v1.items and filtering out?
Also this has the affect of reporting errors while cloud-init is still running because early calls to status may not have crossed the threshold where all cloud-init stages have yet run. Call to status before cloud-init-local even starts would report. Do we want to also check is_running() too before trying to return these errors?   

Also another concern, is if cloud-final.service is ever not part of the systemd (or other) boot targets, we would never create /run/cloud-init/result.json file which would leave us in a state where we probably want to peek at whether the system's desired run level was reached and the init system is 'up/active' too. 

```suggestion
    return [
        f"Failed to start stage: {stage_name}"
        for stage_name in STAGE_NAME.keys()
        if status_v1.get(stage_name, {}).get("start") is None
    ]
```

Reply:
Thanks for the comment and patch. Also, my apologies, I meant to mark this PR as draft as I knew it wasn't quite ready for review yet. I've applied your patch and added an additional unit test.

> Another concern, is if cloud-final.service is ever not part of the systemd (or other) boot targets, we would never create /run/cloud-init/result.json file which would leave us in a state where we probably want to peek at whether the system's desired run level was reached and the init system is 'up/active' too.

I'm open to seeing an implementation of this, but my previous attempt at doing this wasn't it. :wink:

Since that case is outside of what we support, I see it as a future nice to have rather than a need to have.

Reply:
Yep, that concern is just noting a gap in behavior if cloud-final is kicked out of systemd boot goals due to ordering cycle, but, in those events `systemctl status list-units --failed` or `systemctl status cloud-init-main` or `cloud-final` will all show the error condition as well.